pkg/connection/config.go

```diff
@@ -0,0 +1,306 @@
+}
+
+// ReadValidConfig reads the specified configuration
+// and validates it. An error is returned if the Config could not
+// be read or is not valid. Binds to standard Flags().
+func ReadValidConfig(name string) (config Config, err error) {
+	config, err = ReadConfig(name)
+	if err != nil {
+		return
+	}
+	err = config.Validate()
+	return config, err
+}
+
+// ReadConfig loads Config from yaml file matching `name`. If name
+// contains a path, the file will be read from that path, otherwise
+// the path is assumed as: ~/.config/registry.
+// Setting values are prioritized in order from: flags, env vars, file.
+// If name is empty, no file will be loaded and only flags and env vars
+// will be used.
+// Client can call Flags() to get the standard list.
+// env vars: APG_REGISTRY_ADDRESS, APG_REGISTRY_INSECURE, APG_REGISTRY_TOKEN
+// flag names: registry.address, registry.insecure, registry.token
+func ReadConfig(name string) (config Config, err error) {
+	v := viper.New()
+	v.SetConfigType("yaml")
+	if err = v.BindPFlags(Flags); err != nil {
+		return
+	}
+	if err = bindEnvs(v); err != nil {
+		return
+	}
+
+	dir, file := filepath.Split(name)
+	v.SetConfigName(file)
+	if dir != "" {
+		v.AddConfigPath(dir)
+	} else {
+		v.AddConfigPath(ConfigPath)
+	}
+
+	if err = v.ReadInConfig(); err != nil {
+		if _, ok := err.(viper.ConfigFileNotFoundError); ok && name == "" {
+			err = nil
+		} else {
+			return
+		}
+	}
+
+	// add wrapper "registry.xxx " for unmarshal
+	reg := struct {
+		Config Config `mapstructure:"registry"`
+	}{}
+	if err = v.Unmarshal(&reg); err != nil {
+		return
+	}
+	config = reg.Config
+	return
+}
+
+// DeleteConfig deletes a configuration.
+// Will error if active or missing (*os.PathError).
+func DeleteConfig(name string) error {
+	if err := ValidateConfigName(name); err != nil {
+		return err
+	}
+
+	active, err := ActiveConfigName()
+	if err != nil {
+		return err
+	}
+	if name == active {
+		return CannotDeleteActiveError
+	}
+
+	f := filepath.Join(ConfigPath, name)
+	return os.Remove(f)
+}
+
+// returns the config file to use from ~/.config/active_config.
+// Returns "" if active_config is not found.
+func ActiveConfigName() (string, error) {
+	f := filepath.Join(ConfigPath, ActiveConfigPointerFilename)
+	bytes, err := ioutil.ReadFile(f)
+	if errors.Is(err, os.ErrNotExist) {
+		err = nil
+	}
+	return strings.TrimSpace(string(bytes)), err
+}
+
+// binds environment vars to populate config
+func bindEnvs(v *viper.Viper) error {
+	v.AutomaticEnv()
+	v.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
+	v.SetEnvPrefix("APG")
+	bindings := []string{"registry.address", "registry.insecure", "registry.token"}
+	for _, env := range bindings {
+		if err := v.BindEnv(env); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type ValidationError struct {
+	Field      string
+	Validation string
+}
+
+func (e ValidationError) Error() string {
+	return fmt.Sprintf("%s: %s", e.Field, e.Validation)
+}
```

Comment:
This only seems to be used in one place, as `ValidationError{"registry.address", "required"}`. Can we replace it with `var AddressRequiredError = errors.New("registry.address: required")`?

Reply:
It is today, but as this is library code (in `pkg`) the point of this is to have a well-structured format for clients.